8337031: Improvements to CompilationMemoryStatistic

ashu-mehra opened the pull request:
Some minor improvements to CompilationMemoryStatistic. More details are in [JDK-8337031](https://bugs.openjdk.org/browse/JDK-8337031)

Testing:
  test/hotspot/jtreg/compiler/print/CompileCommandPrintMemStat.java
  test/hotspot/jtreg/serviceability/dcmd/compiler/CompilerMemoryStatisticTest.java

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8337031](https://bugs.openjdk.org/browse/JDK-8337031): Improvements to CompilationMemoryStatistic (**Enhancement** - P4)


### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) 🔄 Re-review required (review applies to [008ac6b9](https://git.openjdk.org/jdk/pull/20304/files/008ac6b93d32c32825532d45914434ecb5c4d524))
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20304/head:pull/20304` \
`$ git checkout pull/20304`

Update a local copy of the PR: \
`$ git checkout pull/20304` \
`$ git pull https://git.openjdk.org/jdk.git pull/20304/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20304`

View PR using the GUI difftool: \
`$ git pr show -t 20304`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20304.diff">https://git.openjdk.org/jdk/pull/20304.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20304#issuecomment-2246376684)

ashu-mehra commented:
@tstuefe fyi

vnkozlov commented:
/cc hotspot-compiler

tstuefe commented:
I plan to look at this later this week.

tstuefe commented on src/hotspot/share/memory/arena.cpp:
```diff
@@ -44,6 +44,26 @@ STATIC_ASSERT(is_aligned((int)Chunk::init_size, ARENA_AMALLOC_ALIGNMENT));
 STATIC_ASSERT(is_aligned((int)Chunk::size, ARENA_AMALLOC_ALIGNMENT));
 
+#define ARENA_TAG_STRING_(str) #str
+#define ARENA_TAG_STRING(name, str, desc) ARENA_TAG_STRING_(str),
```
Can you use STR/XSTR in macros.hpp?

tstuefe commented on src/hotspot/share/compiler/compilationMemoryStatistic.cpp:
```diff
@@ -186,10 +200,8 @@ class MemStatEntry : public CHeapObj<mtInternal> {
-  // usage in resource area when total peaked
-  size_t _ra_at_peak;
+  // usage per arena tag when total peaked
+  size_t _tags_size_at_peak[Arena::tag_count()];
```
Can you please make sure Arena::tag_count() evaluates to constexpr? When in doubt, just use the enum value instead.

ashu-mehra replied:
Arena::tag_count() is declared as a constexpr. I wanted to avoid writing `static_cast<int>(Arena::Tag::tag_count)` every time I need tag_count, so I wrapped it in Arena::tag_count() and declared it with constexpr. Is that not sufficient to make it a constexpr?

tstuefe replied:
Okay then, that is fine.

tstuefe commented on src/hotspot/share/compiler/compilationMemoryStatistic.cpp:
```diff
@@ -210,30 +223,45 @@ class MemStatEntry : public CHeapObj<mtInternal> {
-  void set_ra_at_peak(size_t n) { _ra_at_peak = n; }
+  void set_tags_size_at_peak(size_t* tags_size_at_peak, int nelements) {
+    assert(nelements*sizeof(size_t) <= sizeof(_tags_size_at_peak), "overflow check");
+    memcpy(_tags_size_at_peak, tags_size_at_peak, nelements*sizeof(size_t));
```
style, we do blanks between * (n * x, not n*x)

tstuefe reviewed (commented):
Hi Ashu,

generally okay, see remarks.

It seems awkward to have a size_t vector with a defined length, and then having to specify the length as input argument. I'd consider either use the good old style of 

```
void foo(const size_t array[NUM], ...);
```

(using array with a defined size, but careful since in foo sizeof(array) is still just a pointer)

or just write a small wrapper class holding a size_t vector and taking care of the copying.

ashu-mehra commented on src/hotspot/share/compiler/compilationMemoryStatistic.cpp:
```diff
@@ -90,43 +97,50 @@ void ArenaStatCounter::update_c2_node_count() {
+  }
+  if (total != _current) {
+    log_info(compilation, alloc)("WARNING!!! Total does not match current");
+  }
```
This code should not have been there. I forgot to remove it. There is no use of `total` here.

ashu-mehra commented on src/hotspot/share/compiler/compilationMemoryStatistic.cpp:
```diff
@@ -210,30 +223,45 @@ class MemStatEntry : public CHeapObj<mtInternal> {
+    st->print_cr("  " LEGEND_KEY_FMT ": %s", "total", "memory allocated via arenas while compiling");
+    for (int tag = 0; tag < Arena::tag_count(); tag++) {
+      st->print_cr("  " LEGEND_KEY_FMT ": %s", Arena::tag_name[tag], Arena::tag_desc[tag]);
+    }
```
What do you mean by x macro? Do you have an example that shows the use of x macro?

tstuefe replied:
You use them already in your patch.

E.g. 

```
#define XX(name, whatever, desc) st->print_cr("  " LEGEND_KEY_FMT ": " #name #desc
DO_ARENA_TAG(XX)
#undef XX
```

Admittedly, it is not a lot less code than the for loop. Up to you.

ashu-mehra replied:
I will keep the loop if you don't have strong preference for the macro usage here.

ashu-mehra commented:
@tstuefe I have added a patch to address your review comments.

> or just write a small wrapper class holding a size_t vector and taking care of the copying.

Using a wrapper class is a good idea. I have added `ArenaTagsCounter` for that.

tstuefe commented on src/hotspot/share/compiler/compilationMemoryStatistic.hpp:
```diff
@@ -36,30 +36,54 @@ class outputStream;
 class DirectiveSet;
 
+// Helper class to wrap the array of arena tags for easier processing
+class ArenaTagsCounter {
```
Sorry for being a stickler for precise names, but I would like plural for counters here - it is not a single counter, its a series/vector/array of counters.
Any of these work for me: ArenaCountersByTag - ArenaCountersByTagVector - ArenaTagCounterVector - ArenaTagCounters

ashu-mehra replied:
I am pretty bad in naming things, so I welcome these suggestions. I will go with ArenaCountersByTag.

tstuefe reviewed (approved):
Minor naming nit, otherwise good.

ashu-mehra commented:
@tstuefe renamed ArenaTagsCounter to ArenaCountersByTag

vnkozlov reviewed (commented):
Looks reasonable. I submitted our testing to make sure tests passed in our testing.

vnkozlov reviewed (approved):
My testing passed.

ashu-mehra commented:
thanks @tstuefe @vnkozlov for reviewing and testing

ashu-mehra commented:
/integrate
